### PR TITLE
add global config override for increasedCardinality

### PR DESCRIPTION
**Pull request by filintod**
# Description

As part of the work defined in https://github.com/dapr/dapr/issues/7719#issuecomment-2115859125. Adding a global configuration override to help users to set this value in a central place as it is a cross cutting concern, instead of having to change all configurations

## Issue reference

Please reference the issue this PR will close: #7719 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: charts/dapr/charts/dapr_config/templates/dapr_default_config.yaml [hunk 1]
```diff
@@ -15,4 +15,9 @@ spec:
     allowedClockSkew: {{ .Values.global.mtls.allowedClockSkew }}
     controlPlaneTrustDomain: {{ .Values.global.mtls.controlPlaneTrustDomain }}
     sentryAddress: {{ if .Values.global.mtls.sentryAddress }}{{ .Values.global.mtls.sentryAddress }}{{ else }}dapr-sentry.{{ .Release.Namespace }}.svc.cluster.local:443{{ end }}
+  {{ if .Values.global.metrics.http }}
```
Comment (ItalyPaleAle) [inline charts/dapr/charts/dapr_config/templates/dapr_default_config.yaml:18]:
```suggestion
  {{- if .Values.global.metrics.http }}
```

File: charts/dapr/values.yaml [hunk 2]
```diff
@@ -35,6 +35,9 @@ global:
   prometheus:
     enabled: true
     port: 9090
+  metrics: {}
+  # http:
```
Comment (ItalyPaleAle) [inline charts/dapr/values.yaml:39]: Please update the README.md too

File: charts/dapr/values.yaml [hunk 3]
```diff
@@ -35,6 +35,9 @@ global:
   prometheus:
     enabled: true
     port: 9090
+  metrics: {}
```
Comment (ItalyPaleAle) [inline charts/dapr/values.yaml:38]: This needs the `{}` removed

File: pkg/operator/api/configurations.go [hunk 4]
```diff
@@ -25,6 +30,69 @@ import (
...
+
+	if c.Spec.MetricsSpec.HTTP != nil {
+		if c.Spec.MetricSpec == nil {
+			c.Spec.MetricSpec = &configurationapi.MetricSpec{}
+		}
+		c.Spec.MetricSpec.HTTP = c.Spec.MetricsSpec.HTTP
+	}
+}
+
+func (a *apiServer) overrideConfigDefaults(ctx context.Context, cfg *configurationapi.Configuration) error {
```
Comment (ItalyPaleAle) [inline pkg/operator/api/configurations.go:76]: I'm not understanding what this code does. Why do we apply the daprSystem config to sidecars?

Also, what about non-K8s environments?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: pkg/operator/api/configurations.go (hunk 4 above)
Comment (filintod) [reply pkg/operator/api/configurations.go:76]: @ItalyPaleAle The code mainly overrides, if not present on the regular config, config attributes (in this case only one but can be expanded in the future) present on the daprsystem config. Mostly useful for cross cutting concerns where you can define a common config to all.  Now saying this, I'm probably missing the use case of an app without a defined config but for which an operator still want to apply common config setting in the cluster different to dapr internal defaults.

is there something similar to daprsystem config in non-k8s environment? I could look to add to that if there is

File: charts/dapr/values.yaml (hunk 3 above)
Comment (filintod) [reply charts/dapr/values.yaml:38]: was following same pattern as later in this file like
```
  extraVolumes: {}
  #  sentry:
  #  placement:
  #  operator:
  #  injector:
```

Comment (ItalyPaleAle) [reply charts/dapr/values.yaml:38]: I don't believe that's valid YAML syntax

File: pkg/operator/api/configurations.go (hunk 4 above)
Comment (ItalyPaleAle) [reply pkg/operator/api/configurations.go:76]: The daprSystem config is meant to configure the control plane, not individual sidecars.

In non-K8s there's no operator so this code path won't be executed. See the fact that it's not meant to configure individual sidecars.

File: charts/dapr/values.yaml (hunk 3 above)
Comment (filintod) [reply charts/dapr/values.yaml:38]: it is, an empty map in yaml, I can remove it but will have to change the others for consistency

Comment (ItalyPaleAle) [reply charts/dapr/values.yaml:38]:
```
metrics: {}
```

is valid

... but if you uncomment the line below:

```
metrics: {}
   http:
```

is not

Comment (filintod) [reply charts/dapr/values.yaml:38]: this is what I mean of consistency:
```
  metrics: {}
  # http:
  #   increasedCardinality: false # set to true to revert to old behavior by default (dapr < 1.14)
  mtls:
    enabled: true
    workloadCertTTL: 24h
    allowedClockSkew: 15m
    controlPlaneTrustDomain: "cluster.local"
    # If set to true, a bound service account token will be mounted and used to
    # authenticate to Sentry.
    mountSentryVolume: true
    # Used to override `dapr-sentry.{{ .Release.Namespace }}.svc.cluster.local:443`
    #sentryAddress:
  # extraVolumes and extraVolumeMounts are used to mount additional volumes to
  # the Dapr control plane pods. Useful for using alternative authentication
  # credentials to sentry.
  extraVolumes: {}
  #  sentry:
  #  placement:
  #  operator:
  #  injector:
  extraVolumeMounts: {}
  #  sentry:
  #  placement:
  #  operator:
  #  injector:
```

File: pkg/operator/api/configurations.go (hunk 4 above)
Comment (JoshVanL) [reply pkg/operator/api/configurations.go:76]: `mtls.enabled` is an example of an existing configuration today which configures sidecars.

Comment (ItalyPaleAle) [reply pkg/operator/api/configurations.go:76]: Sort-of, as that is still something that also configures the control plane and the entire cluster

We don't have an existing case of using daprSystem for each individual sidecar, also because it doesn't work in standalone mode.

Comment (JoshVanL) [reply pkg/operator/api/configurations.go:76]: `mTLS` is always enabled for control plane components in Kubernetes, so I would argue this setting is only for `daprd`s.

File: charts/dapr/values.yaml (hunk 3 above)
Comment (filintod) [reply charts/dapr/values.yaml:38]: @ItalyPaleAle actually this does not work and will lead to npe :
```
helm template dapr
Error: template: dapr/charts/dapr_config/templates/dapr_default_config.yaml:18:15: executing "dapr/charts/dapr_config/templates/dapr_default_config.yaml" at <.Values.global.metrics.http>: nil pointer evaluating interface {}.http
```

File: charts/dapr/charts/dapr_config/templates/dapr_default_config.yaml (hunk 1 above)
Comment (filintod) [reply charts/dapr/charts/dapr_config/templates/dapr_default_config.yaml]: done

File: charts/dapr/values.yaml (hunk 2 above)
Comment (filintod) [reply charts/dapr/values.yaml:39]: done

File: charts/dapr/values.yaml (hunk 3 above)
Comment (ItalyPaleAle) [reply charts/dapr/values.yaml:38]: What doesn't work?

Comment (filintod) [reply charts/dapr/values.yaml:38]: removing the `{}`

Comment (ItalyPaleAle) [reply charts/dapr/values.yaml:38]: Yes please do that.

Every other option is invalid

```yaml
# This is invalid
metrics:

# This is invalid too
metrics: {}
  http: ...
```

**Review by artursouza — COMMENTED** [review]
LGTM

I think that there is a legit use case for global configuration of sidecars. I do have a concern of this being just a partial implementation and not have all the other settings to also be applied. I am OK to do that as a follow up in 1.15 since the cardinality option is the change with highest priority now.

File: pkg/operator/api/configurations.go (hunk 4 above)
Comment (artursouza) [reply pkg/operator/api/configurations.go:76]: I like the idea to have a central place to change sidecar config without having to manually override each Deployment with annotation.

Now, there is a different approach to this where we could simply allow a Configuration in the "dapr-system" namespace (or whatever the install NS is) to be tagged as the "default" configuration that the sidecar injector will use if Deployment does snot pick one. The pro is that it will be a much simpler change and easy to understand. The disadvantage is that users can only override their app's configuration with an entire new config instead of a merge.

In summary, my question is only if we want to have a "merge" of the default and app's configuration or to have a complete overwrite.

Comment (ItalyPaleAle) [reply pkg/operator/api/configurations.go:76]: This still doesn't solve the problem of Dapr self-hosted. My understanding is that we the project's goal is to maintain feature parity between the two
